utils/s2n_init.c

```diff
@@ -76,8 +84,21 @@ static bool s2n_cleanup_atexit_impl(void)
     return a && b && c;
 }
 
+int s2n_cleanup(void)
+{
+    /* s2n_cleanup is supposed to be called from each thread before exiting,
+     * so ensure that whatever clean ups we have here are thread safe */
+    POSIX_GUARD_RESULT(s2n_rand_cleanup_thread());
+
+    /* If this is the main thread and atexit cleanup is disabled,
+     * perform final cleanup now */
+    if (pthread_self() == main_thread && !atexit_cleanup) {
```

Comment:
https://github.com/aws/s2n-tls/pull/3011#discussion_r693198544 could be an issue for this comparison too